### PR TITLE
Maximum value over a height coordinate

### PR DESCRIPTION
**Pull request by mspelman07**
Related ticket: https://metoffice.atlassian.net/browse/EPPT-188
Acceptance test data: https://github.com/metoppv/improver_test_data/pull/29

Adds in the ability to calculate the maximum value over a height coordinate. This includes the ability to define the bounds of the height coordinate to take the maximum over. This is so we can calculate the maximum relative humidity below 300m.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/utilities/spatial.py
Comment (MoseleyS) [inline improver/utilities/spatial.py:796]: Remove this print statement

File: improver/utilities/spatial.py
Comment (MoseleyS) [inline improver/utilities/spatial.py:813]: The resulting cube has all this in the metadata
- cube.name contains "maximum" and lower and upper height bounds and units.
- cube.coords contains scalar "height" coord with point set to the mid-point and bounds set to the lower and upper height bounds. Coord defines units.
- cube.cell_methods contains info saying this is a maximum over height.

Are we sure we need to change the name too?

File: improver/utilities/spatial.py
Comment (MoseleyS) [inline improver/utilities/spatial.py:775]: Add in that this inherits Iris' meta-data updates to the height coord and to the cell methods.

File: improver/utilities/spatial.py
Comment (MoseleyS) [inline improver/utilities/spatial.py:766]: I'm not sure this is the best place for this new method. All the other methods in `spatial.py` deal with horizontal manipulations. The other possibility is in `cube_manipulation.py`, which contains other collapsing methods - I don't think you want to use `collapsed()` though as it is designed to override Iris' metadata, which are right for this method.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
A few things to think about, but nothing major.

File: improver/utilities/spatial.py
Comment (mspelman07) [reply improver/utilities/spatial.py:813]: I'm not sure. I added the bounds into the name to make sure the name was as accurate as possible to what the diagnostic is. It could just be, for example, "Maximum relative humidity" if we're happy that won't be misleading by implying that it is the maximum over all heights when it's only over a small range of heights. As you mention all the information about the range is included in the other metadata.

It also occurs to me that the lower and upper bounds aren't required to be in metres so I'll need to do something to the name whichever way we choose to go.

File: improver/utilities/spatial.py
Comment (mspelman07) [reply improver/utilities/spatial.py]: Removed

File: improver/utilities/spatial.py
Comment (mspelman07) [reply improver/utilities/spatial.py]: I've added an extra sentence to clarify

File: improver/utilities/spatial.py
Comment (mspelman07) [reply improver/utilities/spatial.py]: I've taken your suggestions and moved the plugin and tests etc. to cube_manipulation.py

File: improver/utilities/spatial.py
Comment (MoseleyS) [reply improver/utilities/spatial.py]: I've asked Bruce for advice here. As for the units of the bounds, they should always be in metres if we are obeying the IMPROVER style guide of always storing data in SI units. I suspect we should therefore enforce SI units here too by stating that the bounds must be in metres and adding a `.convert_units("m")` to the incoming height coord.

**Review by MoseleyS — APPROVED** [review]
There is an outstanding question about meta-data, but otherwise I am happy and this PR can have a second review.

File: improver/utilities/cube_manipulation.py
Comment (MoseleyS) [inline improver/utilities/cube_manipulation.py:758]: Bruce's opinion is that we should not rename the cube as the other meta data describe everything that is needed. Therefore these three lines (and line 736) can be removed. This will change the KGO too.

**Review by MoseleyS — CHANGES_REQUESTED** [review]

File: improver/utilities/cube_manipulation.py
Comment (mspelman07) [reply improver/utilities/cube_manipulation.py]: Metadata has been updated and the acceptance test data recreated.

**Comment by MoseleyS** [comment]
The acceptance test data need fixing (see their PR).

**Review by MoseleyS — APPROVED** [review]

File: improver/utilities/cube_manipulation.py
Comment (brhooper) [inline improver/utilities/cube_manipulation.py:748]: I think it is possible for `cube_subsetted` to be `None` at this point if certain height bounds are input, which will cause an `Attribute_Error` on line 750. 

I think handling of this issue should be added, with an error message explaining that no heights on the input cube are within the given bounds, so the max over height levels cannot be calculated. Tests should also be added for this error.

**Review by brhooper — CHANGES_REQUESTED** [review]
Thanks @mspelman07. This mostly looks fine, I've just added one comment about adding some handling of a possible error.

File: improver/utilities/cube_manipulation.py
Comment (mspelman07) [reply improver/utilities/cube_manipulation.py:752]: You are right this could cause a problem. I've added in an error message and a test for this.

**Review by brhooper — APPROVED** [review]
Thanks for that change @mspelman07, I'm happy with this now

**Pull request merged**
